File: utilities/GoDBTool/src_tool/ConfigFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace src_tool
{
    public static class ELangSql
    {
        public static readonly string PostgresSql = "postgres";
        public static readonly string MySql = "mysql";
        public static readonly string MsSql = "mssql";
        public static readonly string Oracle = "oracle";
    }

    public class ConfigFile
    {
        public TemplateItem Templ_GoLangOrmFile;
        public TemplateItem Templ_GoLangModelFile;

        public TemplateItem Templ_GoLang_SchemaDefItem;
        public TemplateItem Templ_GoLang_SchemaDefItem_Col;
        public TemplateItem Templ_GoLang_ForeignKey;
        
        public string OutputDBContextFile;
        public string Models_Extension;
        public string ModelsOutputDir;
        public string BaseModelName;
        public string ImportPackageModels;
        public string ImportPackageOrm;
        
        public string PackageGenSql;
        public string ConnectionString;
        public string SqlLang = ELangSql.PostgresSql;
        public string Delimeter;
        public string DirJsons;

        public void ParseConfigFile( string pathFile)
        {
            try
            {
                var text = File.ReadAllText(pathFile);
                var lines = text.Split( '\n' );
                for( var iLine = 0; iLine < lines.Length; iLine++ ) 
                {
                    var line1 = lines[ iLine ].Trim();
                    if( line1.StartsWith("#") )
                        continue;

                    var token = getToken( ref line1 );
                    if( token == "$" )
                    {
                        var token1 = getToken( ref line1 );
                        var token2 = getToken( ref line1 );
                        if( token2 == "=" )
                        {
                            var propertyValue = getPropertyValue( line1.Trim());
                            setupSLineProperty( token1, propertyValue );
                        }
                    }else
                    if( token == "@" )
                    {
                        var content = "";
                        iLine++;
                        for( ; iLine < lines.Length; iLine++ ) 
                        {
                            var line2 = lines[ iLine ].Trim();
                            if( line2 == "@#@" )
                                break;
                            content += lines[ iLine ];
                            content += "\r\n";
                        }

                        var token1 = getToken( ref line1 );

                        setupMLineProperty( token1, content );
                    }

                }
            }catch( Exception e )
            {
                Console.WriteLine(e.Message);
            }
            this.checkProps();
        }

        private string getPropertyValue(string val)
        {
            var retValue = "";
            var prevIdx = 0;
            for( var idx = 0; idx < val.Length; )
            {
                idx = val.IndexOf("%", idx);
                if( idx >=0 )
                {
                    retValue += val.Substring( prevIdx, idx);

                    idx = idx+1;
                    var idx2 = val.IndexOf("%", idx);
                    var strProp = val.Substring( idx, idx2-idx);

                    var v = Environment.GetEnvironmentVariable(strProp);
                    if( v != "" && v != null)
                    {
                        retValue += v;
                    }else
                    {
                        retValue += $"%{strProp}%";
                    }
                    prevIdx = idx2+1;
                    idx = prevIdx;
                }else
                    break;
            }

            retValue += val.Substring( prevIdx);
            return retValue;
        }

        private void setupSLineProperty(string token1, string content)
        {
            switch( token1 ) 
            {
                case "OutputDBContextFile": this.OutputDBContextFile = content; break;
                case "Models_Extension": this.Models_Extension = content; break;
                    
                case "ModelsOutputDir":     this.ModelsOutputDir = content; break;
                case "BaseModelName":       this.BaseModelName = content; break;
                case "FullName_PackageModels": this.ImportPackageModels = content; break;
                case "FullName_PackageOrm":    this.ImportPackageOrm = content; break;
                case "FullName_PackageGenSql":       this.PackageGenSql = content; break;
                    
                case "ConnectionString":    this.ConnectionString = content; break;
                case "SqlLang":             this.SqlLang = content; break;
                case "DELIMETER":           this.Delimeter = content; break;
                case "DirJsons":           this.DirJsons = content; break;
                    
                
                    
                default: 
                    Console.WriteLine( $"Not identified token {token1} in single line property ");
                break;
            }
        }
        
        private void setupMLineProperty(string token1, string content)
        {
            switch( token1 ) 
            {
                case "Templ_GoLangOrmFile":             this.Templ_GoLangOrmFile     = new TemplateItem( token1, content ); break;
                case "Templ_GoLangModelFile":           this.Templ_GoLangModelFile   = new TemplateItem( token1, content ); break;
                case "Templ_GoLang_SchemaDefItem":      this.Templ_GoLang_SchemaDefItem     = new TemplateItem( token1, content ); break;
                case "Templ_GoLang_SchemaDefItem_Col":  this.Templ_GoLang_SchemaDefItem_Col = new TemplateItem( token1, content ); break;
                case "Templ_GoLang_ForeignKey":         this.Templ_GoLang_ForeignKey = new TemplateItem( token1, content ); break;

                default: 
                    Console.WriteLine( $"Not identified token {token1} in multiline declaration ");
                break;
            }
        }

        void checkProps()
        {
            var retError = checkAllProps();
            if( retError != "" )
            {
                Console.WriteLine( retError );
            }
        }
        string checkAllProps()
        {
            if( this.OutputDBContextFile == null )
                return $"Missing prop OutputDBContextFile";

            if( this.Models_Extension == null)
                return $"Missing prop Models_Extension";

            if( this.ModelsOutputDir == null)
                return $"Missing prop ModelsOutputDir";

            if( this.BaseModelName == null)
                return $"Missing prop BaseModelName";

            if( this.ImportPackageModels == null)
                return $"Missing prop ImportPackageModels";

            if( this.ImportPackageOrm == null)
                return $"Missing prop ImportPackageOrm";

            if( this.PackageGenSql == null)
                return $"Missing prop PackageGenSql";

            if( this.ConnectionString == null)
                return $"Missing prop ConnectionString";

            if( this.SqlLang == null)
                return $"Missing prop SqlLang";

            if( this.Delimeter == null)
                return $"Missing prop Delimeter";

            if( this.DirJsons == null)
                return $"Missing prop DirJsons";

            if( this.Templ_GoLangOrmFile == null)
                return $"Missing prop Templ_GoLangOrmFile";

            if( this.Templ_GoLangModelFile == null)
                return $"Missing prop Templ_GoLangModelFile";

            if( this.Templ_GoLang_SchemaDefItem == null)
                return $"Missing prop Templ_GoLang_SchemaDefItem";

            if( this.Templ_GoLang_SchemaDefItem_Col == null)
                return $"Missing prop Templ_GoLang_SchemaDefItem_Col";

            if( this.Templ_GoLang_ForeignKey == null)
                return $"Missing prop Templ_GoLang_ForeignKey";

            return "";
        }
        private static string getToken( ref string line)
        {
            var i = 0;
            for( ; i < line.Length; i++ ) 
            {
                var ch = line[ i ];
                if( (ch == '\t'
                 || ch == '\n'
                 || ch == '\r'
                 || ch == ' '))
                {
                }
                else
                    break;
            }

            var content = "";
            for( ; i < line.Length; i++ ) 
            {
                var ch = line[ i ];
                if( ch == '\t'
                 || ch == '\n'
                 || ch == '\r'
                 || ch == ' ')
                {
                    line = line.Substring( i );
                    return content;
                }
                content += ch;
            }
            line = line.Substring( i );
            return content;
        }
        public string UseTemplate(string original, TemplateItem template, Dictionary<string, string>dict)
        { 
            if( template == null) {
                return original;
            }
            return template.ConvertTemplate( dict);
        }
    }
    public class TemplateItem
    {
        public string Name { get; set; }
        public string Text { get; set; }

        public TemplateItem( string name, string text) 
        { 
            Name = name;
            Text = text.Replace("\r\n", "\n");
        }

        public string ConvertTemplate( Dictionary<string, string>dict)
        {
            var text = this.Text;
            var templateName = this.Name;

            foreach( var it in dict )
            {
                var val = $"@@{{{it.Key}}}";

                text = text.Replace( val, it.Value );
            }
            var idx = text.IndexOf("@@{");
            if( idx >= 0)
            {
                var idx2 = text.IndexOf( "}", idx );
                var item = text.Substring( idx, idx2+1-idx);

                Console.WriteLine($"Error : tag {item} is still present in {templateName} ");
            }
            return text;
        }
    }
}

File: library/tools/gomigration/src/ConfigFile.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)



type ConfigFile struct {
	Templ_GoLangOrmFile        *TemplateItem
	Templ_GoLangModelFile      *TemplateItem
	Templ_GoLang_SchemaDefItem *TemplateItem
	Templ_GoLang_SchemaDefItem_Col *TemplateItem
	Templ_GoLang_ForeignKey    *TemplateItem

	OutputDBContextFile  string
	Models_Extension     string
	ModelsOutputDir      string
	BaseModelName        string
	ImportPackageModels  string
	ImportPackageOrm     string
	PackageGenSql        string
	ConnectionString     string
	SqlLang              string
	Delimeter            string
	DirJsons             string
}

func (c *ConfigFile) ParseConfigFile(pathFile string) {
	file, err := os.Open(pathFile)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if strings.HasPrefix(line, "#") {
			continue
		}

		token := getToken(&line)
		if token == "$" {
			token1 := getToken(&line)
			token2 := getToken(&line)
			if token2 == "=" {
				propertyValue := c.getPropertyValue(strings.TrimSpace(line))
				c.setupSLineProperty(token1, propertyValue)
			}
		} else if token == "@" {
			var content strings.Builder
			for scanner.Scan() {
				line2 := strings.TrimSpace(scanner.Text())
				if line2 == "@#@" {
					break
				}
				content.WriteString(line2)
				content.WriteString("\n")
			}

			token1 := getToken(&line)
			c.setupMLineProperty(token1, content.String())
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Println(err)
	}

	c.checkProps()
}

func (c *ConfigFile) getPropertyValue(val string) string {
	retValue := ""
	prevIdx := 0
	for idx := 0; idx < len(val); {
		idx = strings_Index(val, "%", idx)
		if idx >= 0 {
			retValue += val[prevIdx:idx]

			idx = idx + 1
			idx2 := strings_Index(val, "%", idx)
			strProp := val[idx:idx2]

			v := os.Getenv(strProp)
			if v != ""  {
				retValue += v
			} else {
				retValue += fmt.Sprintf("%%%s%%", strProp)
			}
			prevIdx = idx2 + 1
			idx = prevIdx
		} else {
			break
		}
	}

	retValue += val[prevIdx:]
	return retValue
}

func (c *ConfigFile) setupSLineProperty(token1, content string) {
	switch token1 {
	case "OutputDBContextFile":
		c.OutputDBContextFile = content
	case "Models_Extension":
		c.Models_Extension = content
	case "ModelsOutputDir":
		c.ModelsOutputDir = content
	case "BaseModelName":
		c.BaseModelName = content
	case "FullName_PackageModels":
		c.ImportPackageModels = content
	case "FullName_PackageOrm":
		c.ImportPackageOrm = content
	case "FullName_PackageGenSql":
		c.PackageGenSql = content
	case "ConnectionString":
		c.ConnectionString = content
	case "SqlLang":
		c.SqlLang = content
	case "DELIMETER":
		c.Delimeter = content
	case "DirJsons":
		c.DirJsons = content
	default:
		fmt.Printf("Not identified token %s in single line property\n", token1)
	}
}

func (c *ConfigFile) setupMLineProperty(token1, content string) {
	switch token1 {
	case "Templ_GoLangOrmFile":
		c.Templ_GoLangOrmFile = NewTemplateItem(token1, content)
	case "Templ_GoLangModelFile":
		c.Templ_GoLangModelFile = NewTemplateItem(token1, content)
	case "Templ_GoLang_SchemaDefItem":
		c.Templ_GoLang_SchemaDefItem = NewTemplateItem(token1, content)
	case "Templ_GoLang_SchemaDefItem_Col":
		c.Templ_GoLang_SchemaDefItem_Col = NewTemplateItem(token1, content)
	case "Templ_GoLang_ForeignKey":
		c.Templ_GoLang_ForeignKey = NewTemplateItem(token1, content)
	default:
		fmt.Printf("Not identified token %s in multiline declaration\n", token1)
	}
}

func (c *ConfigFile) checkProps() {
	retError := c.checkAllProps()
	if retError != "" {
		fmt.Println(retError)
	}
}

func (c *ConfigFile) checkAllProps() string {
	if c.OutputDBContextFile == "" {
		return "Missing prop OutputDBContextFile"
	}

	if c.Models_Extension == "" {
		return "Missing prop Models_Extension"
	}

	if c.ModelsOutputDir == "" {
		return "Missing prop ModelsOutputDir"
	}

	if c.BaseModelName == "" {
		return "Missing prop BaseModelName"
	}

	if c.ImportPackageModels == "" {
		return "Missing prop ImportPackageModels"
	}

	if c.ImportPackageOrm == "" {
		return "Missing prop ImportPackageOrm"
	}

	if c.PackageGenSql == "" {
		return "Missing prop PackageGenSql"
	}

	if c.ConnectionString == "" {
		return "Missing prop ConnectionString"
	}

	if c.SqlLang == "" {
		return "Missing prop SqlLang"
	}

	if c.Delimeter == "" {
		return "Missing prop Delimeter"
	}

	if c.DirJsons == ""  {
		return "Missing prop DirJsons"
	}

	if c.Templ_GoLangOrmFile == nil {
		return "Missing prop Templ_GoLangOrmFile"
	}

	if c.Templ_GoLangModelFile == nil {
		return "Missing prop Templ_GoLangModelFile"
	}

	if c.Templ_GoLang_SchemaDefItem == nil {
		return "Missing prop Templ_GoLang_SchemaDefItem"
	}

	if c.Templ_GoLang_SchemaDefItem_Col == nil {
		return "Missing prop Templ_GoLang_SchemaDefItem_Col"
	}

	if c.Templ_GoLang_ForeignKey == nil {
		return "Missing prop Templ_GoLang_ForeignKey"
	}

	return ""
}

func getToken(line *string) string {
	i := 0
	for ; i < len(*line); i++ {
		ch := (*line)[i]
		if ch == '\t' || ch == '\n' || ch == '\r' || ch == ' ' {
		} else {
			break
		}
	}

	content := ""
	for ; i < len(*line); i++ {
		ch := (*line)[i]
		if ch == '\t' || ch == '\n' || ch == '\r' || ch == ' ' {
			*line = (*line)[i:]
			return content
		}
		content += string(ch)
	}
	*line = (*line)[i:]
	return content
}

func NewTemplateItem(name, text string) *TemplateItem {
	return &TemplateItem{Name: name, Text: strings.ReplaceAll(text, "\r\n", "\n")}
}

func (c *ConfigFile) UseTemplate(original string, template *TemplateItem, dict map[string]string) string {
	if template == nil {
		return original
	}
	return template.ConvertTemplate(dict)
}




type TemplateItem struct {
	Name string
	Text string
}

func (t *TemplateItem) ConvertTemplate(dict map[string]string) string {
	text := t.Text
	templateName := t.Name

	for key, value := range dict {
		val := fmt.Sprintf("@@{%s}", key)
		text = strings.ReplaceAll(text, val, value)
	}

	idx := strings.Index(text, "@@{")
	if idx >= 0 {
		idx2 := strings_Index( text, "}", idx);
		item := text[idx : idx2+1-idx]

		fmt.Printf("Error: tag %s is still present in %s\n", item, templateName)
	}

	return text
}


